File: ARC/ARC101/MediansOfMediansBIT.java
import java.util.*;

class Main {
    public static void main(String[] args) {
        solve();
    }
    public static void solve(){
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        int a[] = new int[n];
        for(int i=0;i<n;i++) {
            a[i] = sc.nextInt();
        }
        sc.close();
        int maxim = 1000000001;

        int ans = bSearch(0, maxim, a);
        
        System.out.println(ans);      
    }
    
    /*
    条件を満たすxの最大値をにぶたんで求める
    */
    public static int bSearch(int ok,int ng,int[] numbers){
        if(ng == ok + 1){
            return(ok);
        } else {
            int mid = (ok+ng)/2;
            if(judge(mid, numbers)) {
                return(bSearch(mid, ng, numbers));
            } else {
                return(bSearch(ok, mid, numbers));
            }
        }
    }

    public static boolean judge(int x, int[] numbers){
        /*
        int xが条件を満たすかどうか判定する
        条件：x以上のmedian yが半数以上存在する
        */
        int n = numbers.length;
        int cumsum = 0;
        long m = n;
        long intervals = m*(m+1)/2;
        long half = intervals/2;
        long y = 0;
        BIT bit = new BIT(2*n+1);
        bit.add(n+1,1);
        for(int i = 0; i<n; i++){
            if(numbers[i]>=x){
                cumsum++;
            } else {
                cumsum--;
            }
            y += bit.sum(cumsum+n+1);
            bit.add(cumsum+n+1,1);
        }
        return(y >= intervals - half);
    }
}

class BIT {
    private int[] values;
    private int length;

    public BIT(int length) {
        this.length = length;
        values = new int[length+1];
    }

    public void add(int location, int value){
        for(int x = location; x <= this.length; x += x & -x){
            values[x] += value;
        }
    }

    public int sum(int right){
        int ans = 0;
        for(int x = right; x > 0; x -= x & -x){
            ans += values[x];
        }
        return ans;
    }

    public int sum(int left, int right){
        if(left > 1)
            return(sum(right)-sum(left-1));
        else
            return(sum(right));
    }
}

File: ARC/ARC101/MediansOfMediansBIT.py
N = int(input())
a = list(map(int,input().split()))

class BIT():
    def __init__(self,length):
        self.length = length
        self.values = [0 for _ in range(length+1)]
    
    def add(self,location,value):
        x = location
        while x < self.length + 1:
            self.values[x] += value
            x += x & -x
    
    def sum(self,right):
        x = right
        answer = 0
        while x > 0:
            answer += self.values[x]
            x -= x & -x
        return(answer)

def judge(x):
    bit = BIT(2*N + 1)
    bit.add(N+1,1)
    cumsum = 0
    validNum = 0
    interval = N*(N+1)//2
    lowbound = interval - interval//2
    for i in range(N):
        if a[i] >= x:
            cumsum += 1
        else:
            cumsum -= 1
        validNum += bit.sum(cumsum+N+1)
        bit.add(cumsum+N+1,1)
    return(validNum >= lowbound)    
    
def solve(ok,ng):
    """
    解が[ok,ng)に入ることを前提に2分探索をし、judge(x)==Trueとなる最大のxを求める。
    judge(x)の性質は2分探索が可能なものであることを仮定する。
    """
    if ok + 1 == ng:
        return(ok)
    mid = (ok+ng)//2
    if judge(mid):
        return(solve(mid,ng))
    else:
        return(solve(ok,mid))

print(solve(0,10**9+1))